flatsurf/geometry/hyperbolic.py

```diff
@@ -0,0 +1,731 @@
+r"""
+Two dimensional hyperbolic geometry.
+
+EXAMPLES::
+
+    sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+    sage: H = HyperbolicPlane(QQ)
+
+    TODO: More examples.
+
+"""
+######################################################################
+#  This file is part of sage-flatsurf.
+#
+#        Copyright (C) 2022 Julian Rüth
+#                      2022 Sam Freedman
+#
+#  sage-flatsurf is free software: you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation, either version 2 of the License, or
+#  (at your option) any later version.
+#
+#  sage-flatsurf is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with sage-flatsurf. If not, see <https://www.gnu.org/licenses/>.
+######################################################################
+
+from sage.structure.parent import Parent
+from sage.structure.element import Element
+from sage.structure.unique_representation import UniqueRepresentation
+
+
+class HyperbolicPlane(Parent, UniqueRepresentation):
+    r"""
+    The hyperbolic plane over a base ring.
+
+    All objects in the plane must be specified over the given base ring. Note
+    that, in some representations, objects might appear to live in a larger
+    ring. E.g., when specifying a line by giving a midpoint and the square of
+    its radius in the half plane model, then the ideal endpoints of this line
+    might have coordinates in the ring after adjoining a square root.
+
+    The implemented elements of the plane are convex subsets such as (finite
+    and infinite) points, geodesics, closed half planes, and closed convex
+    polygons.
+
+    ALGORITHM:
+
+    We do not use a fixed representation of the hyperbolic plane internally but
+    switch between the Poincaré half plane and the Klein model freely.
+
+    For the Klein model, we use a unit disc centered at (0, 0). The map from
+    the Poincaré half plane sends the imaginary unit `i` to the center at the
+    origin, and sends 0 to (0, -1), 1 to (1, 0), -1 to (-1, 0) and infinity to
+    (0, 1). In other words, the map from the Poincaré model is related to the
+    the Möbius transformation
+
+    .. MATH::
+
+        z \mapsto \frac{z-i}{1 - iz}
+
+    which goes to the Poincaré disc model composed with the map from the
+    Poincaré disc model to the Klein disc model, i.e., the map that, in polar
+    coordinates, maps
+
+    .. MATH::
+
+        (\phi, r)\mapsto \left(\phi, \frac{2r}{1 + r^2}\right).
+
+    When we write this map out explicitly in Euclidean coordinates, we get
+
+    .. MATH::
+
+        (x, y) \mapsto \frac{1}{1 + x^2 + y^2}\left(2x, -1 + x^2 + y^2\right)
```

Comment:
This was verified with Sage?

Reply:
Most of them were done with Sage. The symbolic engine has trouble with square roots, so some of them, I did by hand and then verified that they are inverse to each other on random points.

However, since I did this at 1am on the train, there might be something wrong. I guess we'll notice immediately once we see plots of these.

Reply:
(and thanks for the better wording.)